### ISSUE TITLE
Extremely Slow MMS on AT&T

### ISSUE THREAD

**Issue by klindek** [WhisperSystems/Signal-Android#4474]
Sending and receiving MMS on AT&T takes 20-30 seconds while Google Messenger is nearly instantaneous on a Nexus 6P running Android 6.0. I noticed in the logs that the system was defaulting to the legacy MMS transport rather than Lollipop transport.

**Comment by klindek** [comment]
Also, running the latest Beta published through the Play store.

**Comment by mcginty** [comment]
Those logs would be very helpful to see

**Comment by jabstone** [comment]
I have the same problem and can provide the logs below. I have the problem on both AT&T and Verizon. None of the phones are rooted. The AT&T phone is a Moto G. The Verizon phone is a Galaxy S6. I sent the same picture in each of the tests. The only time Signal worked as expected is when the Verizon phone was not connected to WiFi. Logs for the different test scenarios are below.

Log #1 (AT&T Phone on WiFi) - The following debug log is for an AT&T phone running Android 5.1, connected to WiFi. The MMS message took ~140 seconds to receive after the sending phone had successfully sent the message. The same test took ~30 seconds using Google Messenger.
https://gist.github.com/ed4331d88db38170329b

Log #2 (AT&T Phone on LTE) - This debug log is for the same phone as log #1 except it is not connected to WiFi. The MMS message took ~80 seconds to receive after the sending phone had successfully sent the message. This same test took ~14 seconds on Google Messenger.
https://gist.github.com/07ec055bfefa1391fe01

Log #3 (Verizon Phone on Wifi) - This debug log is for the Verizon phone running Android 5.1.1, connected to WiFi. The MMS message took ~140 seconds to receive after the sending phone had successfully sent the message. The same test took ~10 seconds to receive using Google Messenger.
https://gist.github.com/0fed3adcfc7cbebe4e2e

Log #4 (Verizon Phone on LTE) - This debug log is for the same Verizon phone but not connected to WiFi. The MMS message took ~5 seconds to receive after the sending phone had successfully sent the message. The same test took ~4 seconds to receive using Google Messenger as well.
https://gist.github.com/bed9d057ca446dfea16b

**Comment by jeremymasters** [comment]
Maybe even change the title to reflect Verizon.  I have had this issue with TS/Signal for awhile now but didn't think it was different since I haven't used another sms app for a couple years now.

**Comment by klindek** [comment]
W/IncomingLegacyMmsConnection( 7792): Downloading in MMS mode without proxy... appears to be the culprit. Let me pull my own logs and see what happens.

**Comment by klindek** [comment]
https://gist.github.com/21d8c5fd2cfe71ad2779

**Comment by littleTrex** [comment]
Same issue happens to me. I'm on AT&T netowrk, and MMS messages are slow to send and receive.

https://gist.github.com/9e247a5ce10dd296c68c

**Comment by klindek** [comment]
Problem persists in 3.4.1

**Comment by Himself132** [comment]
Same issue here - sometimes can't even retrieve them after waiting hours (Issue is on wireless or 4G LTE)

A. SilentOS 2.0.1, Android Version 5.1.1
B. BlackPhone 2
C. Signal Version 3.3.3
D. AT&T
E. APN Settings
Name: ATT Phone
APN: Phone
Proxy: Not Set
Port: Not Set
Username: Not Set
Password: Not Set
Server: Not Set
MMSC: http://mmsc.mobile.att.net
MMS Proxy: proxy.mobile.att.net
MMS Port: 80
MCC:  310
MNC: 410
Authentication Type: None
APN Type: default, mms, supl, hipri
APN Protocol: IPv4
APN enable/disable: APN Enabled
Bearer: Unspecified
MVNO Type: None
MVNO Value: Not Set
F. Enhanced 4G LTE Mode, tried it on and off
G. Preferred Network Setting: GSM/WCDMA/LTE

**Comment by privacymatters** [comment]
Also noticing that I am completely unable to send or receive MMS with Orbot (Tor)...SMS will work, but nothing when it comes to MMS. Nexus 6...ATT, Marshmallow

**Comment by mcginty** [comment]
@privacymatters any traffic tunneling usually breaks MMS unless your provider supports MMS requests from The Real Internet™, which is not common.

**Comment by privacymatters** [comment]
Hmmm....no such issues with VPN...just Orbot.

**Comment by Himself132** [comment]
I have a BlackPhone2 and I recently forced Google Play Services to ask for permission to read SMS and MMS, now when I am in Signal I am being prompted to allow Google Play Services access to MMS and SMS whenever it is on a group thread, I have two concerns.

 

One is why in the hell is Google Play Services asking to read my Signal secure messages, and the other is that this may be causing the delays for everyone – unless this is also occurring on iOS?

**Comment by privacymatters** [comment]
As a follow-up, I disabled transparent proxying in Orbot (Tor) so that ONLY the Tor Browser (Orfox) uses it. Tried to send MMS in Signal again and it failed. When I totally turn off Orbot, then MMS goes back to working, albeit excruciatingly slow. Orbot, when not transparent proxying, only uses port 8118. Yet, Signal seems to choke when Orbot is active. What ports is Signal using when unencrypted messaging is taking place?

**Comment by jabstone** [comment]
@mcginty, When I comment out the following code below in the IncomingLegacyMmsConnection class, MMS messages are delivered immediately on Wifi or LTE on Verizon. Commending this code appears to force the phone to connect over the MmsRadio connection and solves the problem for Verizon devices. I don't have my AT&T phone at the moment but I would suspect the AT&T phone won't follow this same code path since it's not a CDMA device. Do you know why the code below exists for CDMA devices...or Verizon in particular since Verizon doesn't allow MMS connections over Wifi? I feel like I might be hijacking this bug since it was originally AT&T specific. Let me know if I should create a new bug specifically for Verizon. Thanks.

//    if (isCdmaDevice()) {
//      Log.w(TAG, "Connecting directly...");
//      try {
//        return retrieve(contentApn, transactionId, false, false);
//      } catch (IOException | ApnUnavailableException e) {
//        Log.w(TAG, e);
//      }
//    }

**Comment by Himself132** [comment]
Why is this in pending information?

**Comment by littleTrex** [comment]
I second Himself132.

Is there any other information we can provide to help with resolving this issue?

**Comment by jabstone** [comment]
Is there a way to view the priority of this bug? I'm not sure if the devs are really still waiting for more information (as the tag implies) or if this is considered low priority and being treated as such. Seems like messaging related bugs would be somewhat high priority for a messaging app. If there's more information we can provide, I would be more than happy to help. Thanks!

**Comment by Himself132** [comment]
I’ll second that…. It would be good to hear – thanks guys we are working on this, or can you please provide us with X.

**Comment by TJC** [comment]
I'm having very similar problems, with MMS taking 30-120 seconds to send or retrieve.
However I'm on the Optus network in Australia.
Sony Xperia Z2, standard firmware.

**Comment by Himself132** [comment]
I myself am afraid that this is a known issue and has been for years.  I had a hiatus from this with Silent Text and came back that was a long long time ago.  They had the same issues then.

**Comment by jjohns71** [comment]
Having this issue as well after recently updating to the Nexus 6P. I am running Android Marshmallow and MMS messages (as part of a group text) will NOT download inline. I see the notification of a new message but the message does not appear in the body of the group text message thread. I do see a new message from each participant with an option to download the message. When I go to the individual contact thread with a notice to download MMS only one message and click download on all the items, only ONE message comes through every 120 seconds or so.

**Comment by bconcio** [comment]
I'm not sure how I can help but I'm having the same exact issue as jjohns71. Nexus 6P on ATT. MMS group chats are taking a while to send/receive. I texted someone with regular SMS as well and they did not receive the texts. If you need any more information I will try to provide what I can.

**Comment by kevin-george** [comment]
Same issue as jjohns71. Here are the debug logs,
https://gist.github.com/9ce183b4fe0f5424e218

**Comment by jjohns71** [comment]
If you look through the debug log posted above by kevin-george I think the issue may have to do with 'APNDatabase' which is using the APN mmsc of 'http://mmsc.cingular.com' as the default AT&T MMS server. 

**TL;DR** I think 'APNDatabase' needs to be updated in the code of Signal to use the 'phone'[1] or 'nxtgenphone'[2] AT&T APN's and not the 'wap.cingular' APN (Please see my explanation below for why I think this is the case).

If you look through the debug logs posted by kevin-george I think the **DELAY** issue we are all having is due to a bunch of connections initially connecting, but ultimately being dropped going to 'http://mmsc.cingular.com'. This MMS server is an old and possibly deprecated APN server that AT&T either provides limited support and/or no support for as best as I can tell.

To further my theory that 'APNDatabase' needs to be updated in the Signal codebase I also removed the 'wap.cingular.com' APN from my device (Settings -> Wireless & Networks -> More -> Cellular Networks -> Access Point Names) and added in the 'phone' APN per the instructions from AT&T's support[1]. I then performed a phone reboot and went back into Signal and initiated another MMS send/receive test and looked at my debug logs and Signal is still defaulting to the 'cingular' APN which doesn't even exist in my phone's APN settings anymore.

Please see the AT&T developer thread[2] with a list of APNs in the reference section below

As a side note, it appears the entire AT&T network is moving towards the 'nxtgenphone' APN but for now, there is not full support (No VoLTE support) for unlocked/unbranded/bring your own device phones (i.e. the entire Google Nexus line) on AT&T[3].

From what I can tell by looking over the debug logs, I think we need to update the Android Signal code which is trying to run unsecured MMS through the Signal APNdatabase which defaults to the old legacy Cingular (the company precursor to AT&T Wireless) servers. Ideally it seems as if the Signal code would be optimal if it just defaulted to the device's APN list. (Maybe this is difficult to implement though?)

If the answer is to update the APN Dataase in Signal's codebase, I think the database of AT&T APNs needs to be updated in Signal to a new default order of:
1) 'phone' 
2) 'nxtgenphone' (Due to this not fully being supported by AT&T on unlocked phones)
3) 'wap.cingular'

I'm not a coder or I would try to figure out what library the 'APNDatabase' code is in and take a look and see if I could just change what appears to be the hardcoded AT&T APN setting in Signal.

Can anyone else having this issue try to confirm/dispute my findings? If it is helpful I can post my own debug logs from a few different captures but I think just looking over kevin-george's is enough to understand what the issue appears to be.

Thanks!

**References**
[1] http://www.att.com/esupport/article.html#!/wireless/KB424489 (Expand android devices list).
[2] http://developer.att.com/technical-library/apns/apn-descriptions-and-characteristics 
[3] https://forums.att.com/t5/Device-Features/Nexus-6P-No-HD-Voice-and-Wifi-Calling/td-p/4346076

**Additional Sources of Discussion You May Find Helpful**
[4] http://forum.xda-developers.com/showthread.php?t=2172110 - xda thread with detailed settings for AT&T APN setup. This lists APN's not officially supplied on the AT&T site so use at your own risk.
[5] https://www.reddit.com/r/nexus6/comments/2myk7x/att_nexus_6_apn/ - Reddit discussion around data and MMS on the Nexus 6p which pointed me to [2].

**Comment by kevin-george** [comment]
Hello,
My problem was resolved after i removed my sim card and inserted it again.
I have attached the logs to show what's different once it started working. I haven't touched any of the APN settings(they're the same as those mentioned by Himself123).

https://gist.github.com/ec9706907e01b5c0682f

**Comment by mcginty** [comment]
Yep, we need to update the APN list.

When I compiled the list that's used right now, I merged the APN lists from Cyanogenmod and Google. Retrospectively, I think that was a mistake since it wasn't clear who "won" and was more likely to yield bad results for popular providers even if it led to more hits for the rare ones.

It might be good for someone to try a build with only the latest Google list (https://android.googlesource.com/device/lge/hammerhead/+/android-6.0.1_r5/apns-full-conf.xml) or latest CM list (https://github.com/CyanogenMod/android_vendor_cm/blob/cm-13.0/prebuilt/common/etc/apns-conf.xml) and see if it behaves better.

**Comment by RichJeanes** [comment]
Will updating the APN list also fix the issue for other carriers? I'm on Verizon and had to switch back to Hangouts for SMS/MMS because group messaging was unusable.

**Comment by mcginty** [comment]
Unfortunately with MMS, there is never a single fix for all issues with different carriers. It may or may not help (but I'm guessing the Verizon issues are more around it being a CDMA network, which is a separate open issue).

**Comment by privacymatters** [comment]
Jake, I'm not **THAT** guy, but I was wondering if you had a timeline for updating the APN list? I would do it myself, but I am sorely lacking in coding. I am _shivering_ with excitement at the possibility of making Signal my **_one_** messaging app!

**Comment by mcginty** [comment]
There's no timeline. I won't be in the US for who knows how long, so can't test this as a fix even if I wanted to.

If anyone would like to test it themselves, though, here's an updated apns.db you can replace the current copy in the repo with (in assets/databases): [apns.db.zip](https://github.com/WhisperSystems/Signal-Android/files/110310/apns.db.zip)

**Comment by jjohns71** [comment]
@mcginty I tested the updated and packaged my own self-signed apk, uninstalled the Google Play version and installed my own version of the APK and this updated APN DB fixes the issues on AT&T! MMS messages now connect SUCCESSFULLY to 'proxy.mobile.att.net' instead of all of the dropped connections to the 'wap.cingular' APN. Any chance we can get this included in the next release?

Thank you for providing the updated APN db! I was having issues trying to create it myself (Self Admitted noob with Android Studio)

Warning: For anyone doing the method I used to test this on their own phone/have it built in before the next release. If you're doing it on your primary device, after uninstalling Signal through Google Play and launching your own apk you will need to re-regiser with Signal and you will get a new key so any contacts that already are on Signal will have to refresh their contact list.

**Comment by ClubWDW** [comment]
@jjohns71 Would you be so kind as to post the apk you created here? If that is allowed?

**Comment by 2-4601** [comment]
@ClubWDW Better option is to join the [beta](https://github.com/WhisperSystems/Signal-Android#joining-the-beta) which has that fix.
You can later leave the beta channel if you want and you will automatically transition back to the stable releases.

**Comment by ClubWDW** [comment]
Hmm. . .I suppose I will have to teach myself to compile. I want no part of Google and am running CM13 with no play store. Unless you know of another way, I don't know what else to do.
